### PR TITLE
Fix event not registered on shutdown

### PR DESCRIPTION
**Pull request by arcusfelis**
This PR addresses "MIM-2316 event_not_registered for c2s_xmpp_element_size_out when stopping the node error".

When we call `mongooseimctl stop` (and there are some websocket connections), we see this error in logs:
`{error,#{what => event_not_registered,labels => #{},event_name => c2s_xmpp_element_size_out} .`

Code in question is this one in mongoose_listener:
```
stop() ->
    Listeners = mongoose_config:get_opt(listen),
    lists:foreach(fun stop_listener/1, Listeners),
    mongoose_instrument:tear_down(instrumentation(Listeners)).
```

So, basically it looks like there are some c2s processes alive on the node after we call tear_down. So, metric c2s_xmpp_element_size_out was there, but not here anymore.

Proposed changes include:
- **Try to reproduce the bug, most likely you will need websocket connections when you stop the server (or stop the listener?).** - Done. Just having a websocket connection is enough.

- **Try to fix the bug by ensuring that there are no c2s processes after the listener is stopped. Make sure no processes could spawn up.** - there was already logic for it, it was using ranch and c2s supervisor. Moving stopping based on the supervisor above before stopping the listeners fixes the issue. But there still a chance that new connections would be established. So, we 1) - disable acceptors. 2) - stop c2s processes. 3) - stop listeners completely. instead.

Global distribution:
- mysql_redis preset was using the same redis for both reg and mim clusters. This means that session from reg was visible in mim cluster over ejabberd_sm_redis. So, eve pid was visible in ejabberd_sm on the mim cluster too. We use different redis database numbers for different clusters now.
- added a test with a websocket client to global distribution too.
- added debug logging into make_error_reply.

Shutdown:
- mongoose_c2s_sup only contains websocket clients sessions (maybe bosh too).
- ranch contains regular XMPP connections.

Shutdown logic now is:
- suspend all ranch acceptors.
- drain connections from both ranch and c2s supervisor - could take some time.
- shotdown listeners and remove instrimentation metrics.

Use `ranch:wait_for_connections(Ref, '==', 0)`, because it is more optimized, than just pulling a list of pids and checking its length in the loop.

Ideas for future:
- can we use websocket process as a c2s process? so we have one process in the system per user. Could be nice savings in Websocket oriented setups.
- Move this code around a bit, so we can have `mongooseimctl suspend_listener` and `mongooseimctl drain` commands, which could be useful for implementing graceful shutdowns in k8s or ansible scripts.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_27](https://circleci.com/gh/esl/MongooseIM/242203) / elasticsearch_and_cassandra_mnesia / 8d032a30fa5dc0bb5db136ecfa46611d19d160da
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4400/242203/elasticsearch_and_cassandra_mnesia.27.1.2-amd64)/ [big](https://esl.github.io/html-zip-reader/PR/4400/242203/elasticsearch_and_cassandra_mnesia.27.1.2-amd64/big.tar.gz//ct_run.test%402c22ef18d8cf.2024-11-19_22.09.47/index.html)
**OK: 471** / **Failed: 2** / **User-skipped: 49** / Auto-skipped: 0 
<details><summary>gdpr_SUITE:retrieve_personal_data:retrieve_personal_data_mam:retrieve_personal_data_mam_cassandra:retrieve_mam_muc</summary>


```erlang
{error,{#{what => invalid_response_code,expected_type => ok,
      response_code => {exit_status,1}},
    [{graphql_helper,assert_response_code,2,
             [{file,"/home/circleci/project/big_tests/tests/graphql_helper.erl"},
              {line,258}]},
     {graphql_helper,get_ok_value,2,
             [{file,"/home/circleci/project/big_tests/tests/graphql_helper.erl"},
              {line,241}]},
     {gdpr_SUITE,request_and_unzip_personal_data,3,
           [{file,"/home/circleci/project/big_tests/tests/gdpr_SUITE.erl"},
            {line,1692}]},
     {gdpr_SUITE,'-retrieve_mam_muc/1-fun-2-',4,
           [{file,"/home/circleci/project/big_tests/tests/gdpr_SUITE.erl"},
            {line,652}]},
     {escalus_story,story,4,
            [{file,"/home/circleci/project/big_tests/_build/default/lib/escalus/src/escalus_story.erl"},
             {line,72}]},
     {test_server,ts_tc,3,[{file,"test_server.erl"},{line,1794}]},
     {test_server,run_test_case_eval1,6,
            [{file,"test_server.erl"},{line,1303}]},
     {test_server,run_test_case_eval,9,
            [{file,"test_server.erl"},{line,1235}]}]}}
```

[Report log](https://esl.github.io/html-zip-reader/PR/4400/242203/elasticsearch_and_cassandra_mnesia.27.1.2-amd64/big.tar.gz//ct_run.test%402c22ef18d8cf.2024-11-19_22.09.47/big_tests.tests.gdpr_SUITE.logs/run.2024-11-19_22.12.00/gdpr_suite.retrieve_mam_muc.html)
</details>
<details><summary>gdpr_SUITE:retrieve_personal_data:retrieve_personal_data_mam:retrieve_personal_data_mam_cassandra:retrieve_mam_muc_private_msg</summary>


```erlang
{error,{#{what => invalid_response_code,expected_type => ok,
      response_code => {exit_status,1}},
    [{graphql_helper,assert_response_code,2,
             [{file,"/home/circleci/project/big_tests/tests/graphql_helper.erl"},
              {line,258}]},
     {graphql_helper,get_ok_value,2,
             [{file,"/home/circleci/project/big_tests/tests/graphql_helper.erl"},
              {line,241}]},
     {gdpr_SUITE,request_and_unzip_personal_data,3,
           [{file,"/home/circleci/project/big_tests/tests/gdpr_SUITE.erl"},
            {line,1692}]},
     {gdpr_SUITE,'-retrieve_mam_muc_private_msg/1-fun-2-',3,
           [{file,"/home/circleci/project/big_tests/tests/gdpr_SUITE.erl"},
            {line,686}]},
     {escalus_story,story,4,
            [{file,"/home/circleci/project/big_tests/_build/default/lib/escalus/src/escalus_story.erl"},
             {line,72}]},
     {test_server,ts_tc,3,[{file,"test_server.erl"},{line,1794}]},
     {test_server,run_test_case_eval1,6,
            [{file,"test_server.erl"},{line,1303}]},
     {test_server,run_test_case_eval,9,
            [{file,"test_server.erl"},{line,1235}]}]}}
```

[Report log](https://esl.github.io/html-zip-reader/PR/4400/242203/elasticsearch_and_cassandra_mnesia.27.1.2-amd64/big.tar.gz//ct_run.test%402c22ef18d8cf.2024-11-19_22.09.47/big_tests.tests.gdpr_SUITE.logs/run.2024-11-19_22.12.00/gdpr_suite.retrieve_mam_muc_private_msg.html)
</details>

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_27](https://circleci.com/gh/esl/MongooseIM/243589) / elasticsearch_and_cassandra_mnesia / eb6c12d08083d17908915da2a2dc1dc21d825667
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4400/243589/elasticsearch_and_cassandra_mnesia.27.1.2-amd64)/ [big](https://esl.github.io/html-zip-reader/PR/4400/243589/elasticsearch_and_cassandra_mnesia.27.1.2-amd64/big.tar.gz//ct_run.test%40456bc3ef9487.2024-11-28_23.46.44/index.html)
**OK: 473** / Failed: 0 / **User-skipped: 49** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_27](https://circleci.com/gh/esl/MongooseIM/243612) / elasticsearch_and_cassandra_mnesia / 0a6f37d6849e11da2d810dc71c84cdd17f81250a
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4400/243612/elasticsearch_and_cassandra_mnesia.27.1.2-amd64)/ [big](https://esl.github.io/html-zip-reader/PR/4400/243612/elasticsearch_and_cassandra_mnesia.27.1.2-amd64/big.tar.gz//ct_run.test%4010c25dda5605.2024-11-29_01.56.53/index.html)
**OK: 473** / Failed: 0 / **User-skipped: 49** / Auto-skipped: 0

**Review by chrzaszcz — COMMENTED** [review]
Looks good, nice to see it fixed 👍 
I added minor comments.

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_27](https://circleci.com/gh/esl/MongooseIM/244301) / elasticsearch_and_cassandra_mnesia / 87f7bf4b05a5630b6087b4c70438c34b3939b997
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4400/244301/elasticsearch_and_cassandra_mnesia.27.1.2-amd64)/ [big](https://esl.github.io/html-zip-reader/PR/4400/244301/elasticsearch_and_cassandra_mnesia.27.1.2-amd64/big.tar.gz//ct_run.test%40af657e8fa436.2024-12-02_12.30.39/index.html)
**OK: 473** / Failed: 0 / **User-skipped: 49** / Auto-skipped: 0

File: src/ejabberd_app.erl [hunk 1]
```diff
@@ -90,18 +87,15 @@ do_start() ->
 %% before shutting down the processes of the application.
 prep_stop(_State) ->
     mongoose_deprecations:stop(),
-    TypedListeners = get_typed_listeners(),
-    suspend_listeners(TypedListeners),
-    StoppedCount = broadcast_c2s_shutdown_sup(),
-    StoppedCount2 = broadcast_c2s_shutdown_to_regular_c2s_connections(TypedListeners),
+    StoppedCount = mongoose_listener:suspend_listeners_and_shutdown_connections(),
```
Comment (chrzaszcz) [inline src/ejabberd_app.erl:90]: Hm... why not call it from `mongoose_listener:stop` if it's going to be called always? Anyway, the current version is ok as well.

**Review by chrzaszcz — APPROVED** [review]
Looks good 👍

File: src/ejabberd_app.erl (hunk 1 above)
Comment (arcusfelis) [reply src/ejabberd_app.erl:90]: It was in the first version I did in the morning.
After I saw a test failing...
So, test is written to mock `mongoose_listener:stop` to test that logic to stop clients works independently from the logic that stops listeners.

So, I've decided to have them separate, to allow to test them separately too.

**Pull request merged**